### PR TITLE
Add support for dyldcache headers up to dyld-1042.1

### PR DESCRIPTION
**Pull request by thestr4ng3r**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Newer dyldcaches have more fields in the header. This change adds heuristics for determining the version and reading the newer fields if available.

**Test plan**

See updated tests

### CODE REVIEW

**Review by XVilka — APPROVED** [review]
Nice heuristics to determine the version

**Comment by XVilka** [comment]
One failed test:
```
[XX] db/formats/dyldcache dyldcache 1042.1 header
RZ_NOPLUGINS=1 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-693079402/rizinorg/rizin/prefix/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc 'i
is
iS
iH~{}
oml
' bins/dyldcache/dyld_shared_cache_arm64-macOS-1042.1
-- stdout
--- expected
+++ actual
@@ -1,173 +1,0 @@
-fd       3
-file     bins/dyldcache/dyld_shared_cache_arm64-macOS-1042.1
-size     0x20000
-humansz  128K
-mode     r-x
-format   dyldcache
-iorw     false
-block    0x100
-type     development
-arch     arm
-cpu      N/A
-baddr    0x180000000
-binsz    0x00020000
-bintype  N/A
-bits     64
-retguard false
-class    dyldcache
-compiler N/A
-dbg_file N/A
-endian   LE
-hdr.csum N/A
-guid     c0a042e3c0db394caa455cd7e767643f
-intrp    N/A
-laddr    0x00000000
-lang     N/A
-machine  arm
-maxopsz  4
-minopsz  4
-os       macOS
-cc       N/A
-pcalign  4
-rpath    N/A
-subsys   xnu
-stripped false
-crypto   false
-havecode true
-va       true
-sanitiz  false
-static   true
-linenum  false
-lsyms    false
-canary   false
-PIE      false
-RELROCS  false
-NX       false
-nth paddr      vaddr       bind   type size lib name                      
---------------------------------------------------------------------------
-0   0x00007fb4 0x180007fb4 GLOBAL FUNC 0        _i_am_libSystem
-0   0x0000bfb0 0x18000bfb0 GLOBAL FUNC 0        _say_hello
-0   0x0000ff38 0x18000ff38 GLOBAL FUNC 0        dyld3::entryVectorForDyld
-1   0x0000ff34 0x18000ff34 GLOBAL FUNC 0        _i_am_libdyld
-0   0x00013fb4 0x180013fb4 GLOBAL FUNC 0        _i_am_libsystem_kernel
-paddr      size vaddr       vsize align perm name                                                 type flags 
--------------------------------------------------------------------------------------------------------------
-0x00007fb4 0x4  0x180007fb4 0x4   0x0   -r-x lib_libSystem.B.dylib.0.__TEXT.__text                     
-0x00007fb8 0x48 0x180007fb8 0x48  0x0   -r-x lib_libSystem.B.dylib.1.__TEXT.__unwind_info              
-0x0000bfb0 0x8  0x18000bfb0 0x8   0x0   -r-x lib_libhello.dylib.0.__TEXT.__text                        
-0x0000bfb8 0x48 0x18000bfb8 0x48  0x0   -r-x lib_libhello.dylib.1.__TEXT.__unwind_info                 
-0x0000ff34 0x4  0x18000ff34 0x4   0x0   -r-x system_libdyld.dylib.0.__TEXT.__text                      
-0x0000ff38 0x80 0x18000ff38 0x80  0x0   -r-x system_libdyld.dylib.1.__TEXT.__const                     
-0x0000ffb8 0x48 0x18000ffb8 0x48  0x0   -r-x system_libdyld.dylib.2.__TEXT.__unwind_info               
-0x00013fb4 0x4  0x180013fb4 0x4   0x0   -r-x system_libsystem_kernel.dylib.0.__TEXT.__text             
-0x00013fb8 0x48 0x180013fb8 0x48  0x0   -r-x system_libsystem_kernel.dylib.1.__TEXT.__unwind_info      
-{
-  "version": "1042.1",
-  "header": {
-    "magic": "dyld_v1   arm64",
-    "mappingOffset": 512,
-    "mappingCount": 2,
-    "imagesOffset": 736,
-    "imagesCount": 4,
-    "dyldBaseAddress": 0,
-    "codeSignatureOffset": 114688,
-    "codeSignatureSize": 16384,
-    "slideInfoOffset": 0,
-    "slideInfoSize": 0,
-    "localSymbolsOffset": 0,
-    "localSymbolsSize": 0,
-    "uuid": "c0a042e3c0db394caa455cd7e767643f",
-    "cacheType": "development",
-    "branchPoolsOffset": 0,
-    "branchPoolsCount": 0,
-    "dyldInCacheMH": 0,
-    "dyldInCacheEntry": 0,
-    "imagesTextOffset": 864,
-    "imagesTextCount": 4,
-    "patchInfoAddr": 6442532976,
-    "patchInfoSize": 264,
-    "otherImageGroupAddrUnused": 0,
-    "otherImageGroupSizeUnused": 0,
-    "progClosuresAddr": 0,
-    "progClosuresSize": 0,
-    "progClosuresTrieAddr": 0,
-    "progClosuresTrieSize": 0,
-    "platform": 1,
-    "formatVersion": 0,
-    "dylibsExpectedOnDisk": 0,
-    "simulator": 0,
-    "locallyBuiltCache": 1,
-    "builtFromChainedFixups": 0,
-    "padding": 0,
-    "sharedRegionStart": 6442450944,
-    "sharedRegionSize": 131072,
-    "maxSlide": 2147401728,
-    "dylibsImageArrayAddr": 0,
-    "dylibsImageArraySize": 0,
-    "dylibsTrieAddr": 6442532864,
-    "dylibsTrieSize": 112,
-    "otherImageArrayAddr": 0,
-    "otherImageArraySize": 0,
-    "otherTrieAddr": 0,
-    "otherTrieSize": 0,
-    "mappingWithSlideOffset": 624,
-    "mappingWithSlideCount": 2,
-    "dylibsPBLStateArrayAddrUnused": 0,
-    "dylibsPBLSetAddr": 6442533240,
-    "programsPBLSetPoolAddr": 6442534280,
-    "programsPBLSetPoolSize": 16384,
-    "programTrieAddr": 6442550664,
-    "programTrieSize": 88,
-    "osVersion": 851968,
-    "altPlatform": 0,
-    "altOsVersion": 0,
-    "swiftOptsOffset": 0,
-    "swiftOptsSize": 0,
-    "subCacheArrayOffset": 992,
-    "subCacheArrayCount": 0,
-    "symbolFileUUID": "00000000000000000000000000000000",
-    "rosettaReadOnlyAddr": 0,
-    "rosettaReadOnlySize": 0,
-    "rosettaReadWriteAddr": 0,
-    "rosettaReadWriteSize": 0,
-    "cacheSubType": 0,
-    "objcOptsOffset": 0,
-    "objcOptsSize": 0,
-    "cacheAtlasOffset": 0,
-    "cacheAtlasSize": 0,
-    "dynamicDataOffset": 114688,
-    "dynamicDataMaxSize": 16384
-  },
-  "images": [
-    {
-      "uuid": "e0a310e1c61e3b289c5df43f04e71a44",
-      "address": 6442467328,
-      "textSegmentSize": 16384,
-      "path": "/usr/lib/libSystem.B.dylib",
-      "name": "libSystem.B.dylib"
-    },
-    {
-      "uuid": "f2c8e54bd02937e69928edd8b4bbec63",
-      "address": 6442483712,
-      "textSegmentSize": 16384,
-      "path": "/usr/lib/libhello.dylib",
-      "name": "libhello.dylib"
-    },
-    {
-      "uuid": "439351ae77ca308ebf1da32729a79709",
-      "address": 6442500096,
-      "textSegmentSize": 16384,
-      "path": "/usr/lib/system/libdyld.dylib",
-      "name": "libdyld.dylib"
-    },
-    {
-      "uuid": "4ce071b5dd373397aca699d517bdb6e7",
-      "address": 6442516480,
-      "textSegmentSize": 16384,
-      "path": "/usr/lib/system/libsystem_kernel.dylib",
-      "name": "libsystem_kernel.dylib"
-    }
-  ]
-}
- 1 fd: 3 +0x00000000 0x180000000 - 0x180013fff r-x fmap.cache_map.0
- 2 fd: 3 +0x00014000 0x180014000 - 0x18001bfff r-- fmap.cache_map.1

-- stderr
ERROR: [r] Cannot open 'bins/dyldcache/dyld_shared_cache_arm64-macOS-1042.1'

-- exit status: 1
```

File: librz/bin/format/mach0/dyldcache.c [hunk 1]
```diff
@@ -49,33 +52,173 @@ static void free_bin(RzDyldBinImage *bin) {
...
+	}
+
+	// Size of the header was continuously expanded during versions as newer fields got added.
+	// There is no dedicated header size stored in the file, but in practice the mappingOffset is
+	// directly after the header, so we use this as an indicator on how far we may read.
+	ut64 hdr_end_offset = offset + hdr->mappingOffset;
+#define READ_OR_FINISH(bits, dst) \
+	do { \
+		if (cur + bits / 8 > hdr_end_offset || !rz_buf_read_le##bits##_offset(cache_buf, &cur, dst)) { \
+			goto finish; \
```
Comment (wargio) [inline librz/bin/format/mach0/dyldcache.c:97]: shouldn't this be `fail` ?

File: librz/bin/format/mach0/dyldcache.c [hunk 2]
```diff
@@ -49,33 +52,173 @@ static void free_bin(RzDyldBinImage *bin) {
...
+	READ_OR_FINISH(32, &hdr->altPlatform);
+	READ_OR_FINISH(32, &hdr->altOsVersion);
+	READ_OR_FINISH(64, &hdr->swiftOptsOffset);
+	READ_OR_FINISH(64, &hdr->swiftOptsSize);
+	READ_OR_FINISH(32, &hdr->subCacheArrayOffset);
+	READ_OR_FINISH(32, &hdr->subCacheArrayCount);
+
+	if (cur + sizeof(hdr->symbolFileUUID) > hdr_end_offset ||
+		!rz_buf_read_offset(cache_buf, &cur, hdr->symbolFileUUID, sizeof(hdr->symbolFileUUID))) {
+		goto finish;
```
Comment (wargio) [inline librz/bin/format/mach0/dyldcache.c:154]: same here

**Review by wargio — APPROVED** [review]

**Pull request merged**